src/datanode/src/region_server.rs

```diff
@@ -306,48 +348,109 @@ impl RegionServerInner {
+            },
         };
+
         let engine_type = engine.name();
 
-        let result = engine
+        // Sets the RegionStatus before the operation.
+        match region_change {
+            RegionChange::Register(_) => {
+                self.region_map.insert(
+                    region_id,
+                    RegionEngineWithStatus::Registering(engine.clone()),
+                );
+            }
+            RegionChange::Deregisters => {
+                self.region_map.insert(
+                    region_id,
+                    RegionEngineWithStatus::Deregistering(engine.clone()),
+                );
+            }
+            _ => {}
+        }
+
+        match engine
             .handle_request(region_id, request)
             .trace(info_span!(
                 "RegionEngine::handle_region_request",
                 engine_type
             ))
             .await
-            .with_context(|_| HandleRegionRequestSnafu { region_id })?;
-
-        match region_change {
-            RegionChange::None => {}
-            RegionChange::Register(_) => {
-                info!("Region {region_id} is registered to engine {engine_type}");
-                self.region_map.insert(region_id, engine);
-                self.event_listener.on_region_registered(region_id);
+            .with_context(|_| HandleRegionRequestSnafu { region_id })
+        {
+            Ok(result) => {
+                match region_change {
+                    RegionChange::None => {}
+                    RegionChange::Register(_) => {
+                        info!("Region {region_id} is registered to engine {engine_type}");
+                        self.region_map
+                            .insert(region_id, RegionEngineWithStatus::Ready(engine));
+                        self.event_listener.on_region_registered(region_id);
+                    }
+                    RegionChange::Deregisters => {
+                        info!("Region {region_id} is deregistered from engine {engine_type}");
+                        self.region_map
+                            .remove(&region_id)
+                            .map(|(id, engine)| engine.set_writable(id, false));
+                        self.event_listener.on_region_deregistered(region_id);
+                    }
+                }
+                Ok(result)
             }
-            RegionChange::Deregisters => {
-                info!("Region {region_id} is deregistered from engine {engine_type}");
-                self.region_map
-                    .remove(&region_id)
-                    .map(|(id, engine)| engine.set_writable(id, false));
-                self.event_listener.on_region_deregistered(region_id);
+            Err(err) => {
+                // Removes the RegionStatus if the operation fails.
+                match region_change {
+                    RegionChange::None => {}
+                    RegionChange::Register(_) | RegionChange::Deregisters => {
```

Comment:
Can it simplify the code if we add an `oneshot` channel to `RegionOpenRequest` to indicate the region open result, then handle the result separately, avoiding this long and complex "match"?

Reply:
I decided to move them into separate methods.